File: 817.linked-list-components.py
#
# @lc app=leetcode id=817 lang=python3
#
# [817] Linked List Components
#

# @lc code=start
# Definition for singly-linked list.
from typing import Optional, List
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def numComponents(self, head: Optional[ListNode], nums: List[int]) -> int:
        nums = set(nums)
        ans = 0
        while head:
            if head.val in nums:
                while head and head.val in nums:
                    head = head.next
                ans += 1
            else:
                head = head.next
        return ans
        
# @lc code=end

File: 817.linked-list-components.java
/*
 * @lc app=leetcode id=817 lang=java
 *
 * [817] Linked List Components
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
import java.util.HashSet;
class Solution {
    public int numComponents(ListNode head, int[] nums) {
        int ans = 0;
        HashSet<Integer> set = new HashSet<>();
        for(int num: nums) set.add(num);
        while(head != null){
            if(set.contains(head.val)){
                while(head != null  && set.contains(head.val)){
                    head = head.next;
                }
                ans++;
            }
            else{
                head = head.next;
            }
        }
        return ans; 

    }
}
// @lc code=end

